File: CS6160_Cryptology/Assignment 2/ProgHW2B/aesLongKeyGen16.c
#include<stdio.h>
#include<stdint.h>

/*This function expands a 16-bit key into a 128-bit key*/
void expandKey(uint8_t* longKey, const uint8_t* shortKey);

/*Test code to illustrate the key expansion*/

int main()
{
    uint8_t shortKey[2]={0xa1,0xb2}; //This is only an example.
    uint8_t longKey[16];

    expandKey(longKey,shortKey);

    for (int i=0;i<16;i++)
        printf("%x ",longKey[i]);

    return 0;
}


void expandKey(uint8_t* longKey, const uint8_t* shortKey)
{
    longKey[0]=shortKey[0];
    longKey[1]=0x94;
    longKey[2]=shortKey[1];
    longKey[3]=0x5a;
    longKey[9]=shortKey[0]^shortKey[1];
    longKey[10]=0xe7;

    for (int i=4;i<=15;i++)
    {
        if ((i==9)||(i==10)) continue;
        longKey[i]=(longKey[i-1]+longKey[i-4])%257;
        if (longKey[i]==256)
            longKey[i]==0;
    }
}

File: CS6160_Cryptology/Assignment 2/ProgHW2B/aesLongKeyGen16.py
#Expands a 2-byte input into 16 bytes
def expandKey(shortKey):    
    shortKeyval1=shortKey[0]
    shortKeyval2=shortKey[1]    
    
    ByteA=shortKeyval1.to_bytes(1,"big")
    ByteB=shortKeyval2.to_bytes(1,"big")
    ByteC=(shortKeyval1^shortKeyval2).to_bytes(1,"big")
    hexByte1=0x94
    Byte1=hexByte1.to_bytes(1,"big")
    hexByte2=0x5a
    Byte2=hexByte2.to_bytes(1,"big")
    hexByte3=0xe7
    Byte3=hexByte3.to_bytes(1,"big")
    
    longKey=bytearray(ByteA)    
    longKey.extend(Byte1)
    longKey.extend(ByteB)    
    longKey.extend(Byte2)
    
    for i in range(4,9):        
        hexByte=(longKey[i-1]+longKey[i-4])%257
        if (hexByte==256):
            hexByte=0
        Byte=hexByte.to_bytes(1,"big")              
        longKey.extend(Byte)
    longKey.extend(ByteC)
    longKey.extend(Byte3)
    for i in range(11,16):
        hexByte=(longKey[i-1]+longKey[i-4])%257
        if (hexByte==256):
            hexByte=0
        Byte=hexByte.to_bytes(1,"big")              
        longKey.extend(Byte)    
    
    return longKey

#Test key expansion
'''shortKey=bytearray([0xa1,0xb2,0xc3])
longKey1=expandKey(shortKey)
print(longKey1)'''